File: python/infinicore/ops/bucketize.py
import infinicore
from infinicore.lib import _infinicore
from infinicore.tensor import Tensor


def bucketize(
    input: Tensor,
    boundaries: Tensor,
    *,
    out=None,
    right=False,
) -> Tensor:
    r"""Apply the bucketize function."""

    if infinicore.use_ntops and input.device.type in ("cuda", "musa"):
        return infinicore.ntops.torch.bucketize(input, boundaries, out=out, right=right)

    if out is None:
        return Tensor(
            _infinicore.bucketize(
                input._underlying, boundaries._underlying, bool(right)
            )
        )

    _infinicore.bucketize_(
        input._underlying, boundaries._underlying, out._underlying, bool(right)
    )
    return out

File: src/infinicore/ops/bucketize/bucketize.cc
#include "infinicore/ops/bucketize.hpp"
#include "../../utils.hpp"

namespace infinicore::op {

common::OpDispatcher<Bucketize::schema> &Bucketize::dispatcher() {
    static common::OpDispatcher<Bucketize::schema> dispatcher_;
    return dispatcher_;
};

void Bucketize::execute(Tensor input, Tensor boundaries, Tensor output, bool right) {
    infinicore::context::setDevice(input->device());
    dispatcher().lookup(input->device().getType())(input, boundaries, output, right);
}

Tensor bucketize(Tensor input, Tensor boundaries, bool right) {
    auto output = Tensor::empty(input->shape(), DataType::I64, input->device());
    bucketize_(input, boundaries, output, right);
    return output;
}

void bucketize_(Tensor input, Tensor boundaries, Tensor output, bool right) {
    Bucketize::execute(input, boundaries, output, right);
}

} // namespace infinicore::op